### PR TITLE
Fix unsafe URLs in audit log resulting from domain blocks

### LINKED ISSUE #25842: Audit log for domain blocks links to potentially dangerous domains
### Steps to reproduce the problem

1. Perform a domain-level moderation action of any kind.
2. Look at the audit log for that event.
3. Observe (but be careful with clicking) the link destination for the domain you moderated.

### Expected behaviour

The moderated domain should link to the page for that domain inside the instance's admin interface, like users do for their respective user pages.

### Actual behaviour

The moderated domain links directly to the domain in question, potentially sending you off to an unsafe site you've just moderated.

### Detailed description

This is bad behavior for two reasons: 

1. Safety - moderated domains are often moderated for good reason, and this could end up with an admin visiting them accidentally (and potentially with a referer of the instance's admin interface, too).
2. Consistency - other entities (specifically usernames) in the audit log link back into the admin interface, and this is a totally logical/expected/safe behavior. Having different entities link to different places is jarring and accentuates the danger.

### Specifications

Mastodon 4.1.4

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
Fixes #25842

We don't currently have a "show" page for email domain blocks, so I can only link to the index page, and the rows don't contain any identifiers so I can't even do a relative link within the page, so just linking to the index page will have to do.

I noticed this after testing some stuff in the audit log for domain blocks and accidentally opening a horrific domain.

### CODE REVIEW

**Comment by ThisIsMissEm** [comment]
There doesn't currently appear to be any test coverage on `Admin::ActionLogsHelper` but I don't think we should prevent that from landing a security/safety fix.

**Comment by ThisIsMissEm** [comment]
I would definitely be in favour of some assistance with resolving the failure in `./spec/controllers/admin/action_logs_controller_spec.rb` as that test case is built on top of `orphaned_logs` which results in things like DomainBlocks and UnavailableDomains not actually being full records, as expected by the code.

Feels like we should actually fabricate one of each entity specifically for testing the audit log & the reference is correct?

**Comment by ClearlyClaire** [comment]
Linking to a moderation page rather than the original domain seems worthwhile, but this has several drawbacks, at least in its current implementation:
- action logs for domain blocks/allows there exist no record of will link to a 404, which is confusing and not useful
- action logs for email blocks will link to the general section on blocks, which is confusing and not really useful either

If the worry is accidentally navigating to something awful, maybe not linking, or adding a confirmation dialog, would make sense instead?

As for the test failure, it is because action logs can be about records that are since deleted/revoked: undone domain blocks, deleted accounts and so on. In your case, you're dereferencing `log.target.domain`, which is unsafe for e.g. domain blocks that have since been revoked. `human_identifier` will contain the domain, even if the domain block gets deleted, so it should be used instead.

**Comment by ClearlyClaire** [comment]
Opened https://github.com/mastodon/mastodon/pull/27149 as an alternative, but now that I think about it, it still allows “open in new tab” behavior without confirmation, so it only partially solves the issue.

**Comment by ThisIsMissEm** [comment]
> Opened #27149 as an alternative, but now that I think about it, it still allows “open in new tab” behavior without confirmation, so it only partially solves the issue.

Yeah, I don't think allowing you to leave the instance from the audit logs is a good idea at all. That should be a very deliberate action, and not something that is mixed in with other links that are internal to the moderation & management tools.

As for deleted records, I was hesitant to use just `log.human_identifier` in the URL creation, because that's not strictly the object that the record is pointing to.

Perhaps we should just store a `deleted_at` for domain_blocks & domain_allows, rather than actually deleting the records?

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire for IP Blocks, I think we need a show/edit page for them. Which would resolve the issue of linking to the index page, but I consider that out-of-scope of this fix for now.

**Comment by ThisIsMissEm** [comment]
> I don't think that's a good idea, keeping the blocks around is not useful, and would require changes to queries and indexes.

Wouldn't this just be a matter of making use of the `include Discard::Model` functionality? i.e., installing that for domain blocks? Then when creating new you'd be doing a "find including discarded or create", essentially?

**Comment by ThisIsMissEm** [comment]
For the `Instance` model, could we make use of `to_log_route_param` there (just aliasing to `to_param` instead of using `human_identifier`? That way we could do: `admin_instance_path(log.route_param)` instead, which feels much nicer?

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire even when using `log.human_identifier` the tests fail because the `target` is synthetic and doesn't actually have a `target.to_log_human_identifier` method, as we're just passing `target_type` and `target_id` instead of an actual `target` model, which also isn't how the action log items are typically created: https://github.com/mastodon/mastodon/blob/0396acf39ea902688374fac65fa7ef5dc4c05512/app/controllers/concerns/accountable_concern.rb

In other words, I think the test for `Admin::ActionLogsController` is poorly written, as it's exercising the code in a manner that is completely divorced from how `Admin::ActionLog` models are typically created.

**Comment by ClearlyClaire** [comment]
Sorry I missed this.

The tests are specifically for handling “orphaned” logs as can be found in the wild. Indeed https://github.com/mastodon/mastodon/pull/19040 unfortunately introduced a migration that would outright lose some vital information about some action log types. So action logs with an unset `human_identifier` do exist in the wild and must be accounted for.

Before your PR, they'd lead to broken `https://` links, with your PR they lead to a crash. Neither is particularly satisfying, but the crash is worse as it prevents seeing the audit log at all.

**Comment by ThisIsMissEm** [comment]
Aah, okay; could we do a migration to fix these orphaned logs? Or can more be created?

I think these would only occur if the instance was deleted, and for Domain logs, I don't think we're even using the target property, just the `human_identifier` and `route_param`.

My big concern here is that we're currently using the `human_identifier` for constructing route parameters, which just feels dangerous.

@ClearlyClaire I think it'd be both wise to test for broken data (e.g., orphaned logs) and to test for a regularly expected log, such that we're testing both happy and sad paths?

i did start creating a migration to backfill the `route_param` from `human_identifier` too, to prevent future issues if we change the human identifier in some way that's incompatible with routing.

**Comment by ClearlyClaire** [comment]
> Aah, okay; could we do a migration to fix these orphaned logs? Or can more be created?

I already added migrations to backfill what we can do (#22091) but some information is just irremediably lost.

> I think these would only occur if the instance was deleted, and for Domain logs, I don't think we're even using the target property, just the `human_identifier` and `route_param`.

It would occur for instances blocked then unblocked before Mastodon 4.0.

> My big concern here is that we're currently using the `human_identifier` for constructing route parameters, which just feels dangerous.

Fair. I'm not sure why it hasn't been implemented for all action log types. I guess to reduce data duplication?

> @ClearlyClaire I think it'd be both wise to test for broken data (e.g., orphaned logs) and to test for a regularly expected log, such that we're testing both happy and sad paths?

Makes sense.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire @renchap would we be comfortable landing this change without tests for 4.3?

**Comment by renchap** [comment]
I will defer on @ClearlyClaire for this

**Comment by mjankowski** [comment]
Might make sense to add `rel=noreferrer` to the outbound links as well?

**Comment by ThisIsMissEm** [comment]
> Might make sense to add `rel=noreferrer` to the outbound links as well?

This should remove all outbound links in this view iirc

**Comment by ClearlyClaire** [comment]
> @ClearlyClaire @renchap would we be comfortable landing this change without tests for 4.3?

Do you mean outright removing the existing tests that are now failing?

> Might make sense to add `rel=noreferrer` to the outbound links as well?

We already do `Referrer-Policy: same-origin` on all pages, so outbound links should not have referrers.

**Comment by ThisIsMissEm** [comment]
> > @ClearlyClaire @renchap would we be comfortable landing this change without tests for 4.3?
> 
> Do you mean outright removing the existing tests that are now failing?

i can't even remember what tests were failing here, I know a different branch it was the ones that were testing corrupted or bad state in the database, and there wasn't really a way to fix those iirc?

**Comment by ClearlyClaire** [comment]
> > > @ClearlyClaire @renchap would we be comfortable landing this change without tests for 4.3?
> > 
> > 
> > Do you mean outright removing the existing tests that are now failing?
> 
> i can't even remember what tests were failing here, I know a different branch it was the ones that were testing corrupted or bad state in the database, and there wasn't really a way to fix those iirc?

The test failure is:
```
  1) Admin::ActionLogsController GET #index returns 200
     Failure/Error: link_to log.human_identifier, admin_instance_path(log.human_identifier)
     
     ActionView::Template::Error:
       No route matches {:action=>"show", :controller=>"admin/instances", :format=>"html", :id=>nil}, possible unmatched constraints: [:id]
       Did you mean?  admin_instance_url
     # ./app/helpers/admin/action_logs_helper.rb:19:in `log_target'
     # ./app/views/admin/action_logs/_action_log.html.haml:7:in `_app_views_admin_action_logs__action_log_html_haml__3572464370620890285_976080'
     # ./app/views/admin/action_logs/index.html.haml:23:in `_app_views_admin_action_logs_index_html_haml___796091925821478887_975700'
     # ./app/controllers/concerns/localized.rb:11:in `set_locale'
     # ./spec/controllers/admin/action_logs_controller_spec.rb:20:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:128:in `block (2 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # ActionController::UrlGenerationError:
     #   No route matches {:action=>"show", :controller=>"admin/instances", :format=>"html", :id=>nil}, possible unmatched constraints: [:id]
     #   Did you mean?  admin_instance_url
     #   ./app/helpers/admin/action_logs_helper.rb:19:in `log_target'
```

It is indeed about “orphaned” log entries, for which information has been irremediably lost. However, this test failure indicates that the admin log page would outright error out if the page includes an orphaned log entry.

**Comment by ThisIsMissEm** [comment]
> > > > @ClearlyClaire @renchap would we be comfortable landing this change without tests for 4.3?
> > > 
> > > 
> > > Do you mean outright removing the existing tests that are now failing?
> > 
> > 
> > i can't even remember what tests were failing here, I know a different branch it was the ones that were testing corrupted or bad state in the database, and there wasn't really a way to fix those iirc?
> 
> The test failure is:
> 
> ```
>   1) Admin::ActionLogsController GET #index returns 200
>      Failure/Error: link_to log.human_identifier, admin_instance_path(log.human_identifier)
>      
>      ActionView::Template::Error:
>        No route matches {:action=>"show", :controller=>"admin/instances", :format=>"html", :id=>nil}, possible unmatched constraints: [:id]
>        Did you mean?  admin_instance_url
>      # ./app/helpers/admin/action_logs_helper.rb:19:in `log_target'
>      # ./app/views/admin/action_logs/_action_log.html.haml:7:in `_app_views_admin_action_logs__action_log_html_haml__3572464370620890285_976080'
>      # ./app/views/admin/action_logs/index.html.haml:23:in `_app_views_admin_action_logs_index_html_haml___796091925821478887_975700'
>      # ./app/controllers/concerns/localized.rb:11:in `set_locale'
>      # ./spec/controllers/admin/action_logs_controller_spec.rb:20:in `block (3 levels) in <top (required)>'
>      # ./spec/rails_helper.rb:128:in `block (2 levels) in <top (required)>'
>      # ------------------
>      # --- Caused by: ---
>      # ActionController::UrlGenerationError:
>      #   No route matches {:action=>"show", :controller=>"admin/instances", :format=>"html", :id=>nil}, possible unmatched constraints: [:id]
>      #   Did you mean?  admin_instance_url
>      #   ./app/helpers/admin/action_logs_helper.rb:19:in `log_target'
> ```
> 
> It is indeed about “orphaned” log entries, for which information has been irremediably lost. However, this test failure indicates that the admin log page would outright error out if the page includes an orphaned log entry.

ah, yeah, that's what I was thinking of, and I don't know the appropriate solution here.

**Comment by ClearlyClaire** [comment]
> ah, yeah, that's what I was thinking of, and I don't know the appropriate solution here.

We just need to handle the special case of `human_identifier` being `nil`. Do you want me to take over this PR?

**Review by ClearlyClaire — APPROVED** [review]

**Review by renchap — APPROVED** [review]
